### PR TITLE
fix(treesitter): update folds in all relevant windows

### PR DESCRIPTION
**Pull request by tomtomjhj**
<details>
<summary>outdated</summary>
Problem: :diffput/get opens diff folds if treesitter foldexpr is used.

Solution: Update/open folds only if foldmethod is still 'expr'.
</details>

Problem: When using treesitter foldexpr,
* :diffput/get open diff folds, and
* folds are not updated in other windows that contain the updated
  buffer.

Solution: Update folds in all windows that contain the updated buffer
and use expr foldmethod.

### CODE REVIEW

File: runtime/lua/vim/treesitter/_fold.lua [hunk 1]
```diff
@@ -233,6 +233,10 @@ local M = {}
 local foldinfos = {}
 
 local function recompute_folds()
+  if vim.wo.foldmethod ~= 'expr' then
```
Comment (lewis6991) [inline runtime/lua/vim/treesitter/_fold.lua:236]: Can you add a comment saying that foldexpr could have changed since this function is scheduled.

Comment (lewis6991) [reply runtime/lua/vim/treesitter/_fold.lua:236]: Does the callback to the autocmd below also need the same guard?

Comment (tomtomjhj) [reply runtime/lua/vim/treesitter/_fold.lua:236]: > Does the callback to the autocmd below also need the same guard?

i think so. will add soon.

btw there is another case that could go wrong: when text changes in a buffer that is not shown in the current window.

**Comment by tomtomjhj** [comment]
https://github.com/neovim/neovim/blob/2d40f5e843a99a67efd55304d17ccd48d47d9013/runtime/lua/vim/treesitter/_fold.lua#L264-L265

@lewis6991 is this still the case? I think this could've been fixed by https://github.com/neovim/neovim/pull/24167. If so, we can remove `schedule_if_loaded` workaround altogether.

**Comment by lewis6991** [comment]
Whenever you defer code with `vim.schedule` a buffer can be removed for any reason so you always need to check if it still exists.

**Comment by tomtomjhj** [comment]
What I mean was that we don't need to defer `get_folds_levels`, because it now does parsing itself. I think "use the old buffer state" might have been caused by not parsing.

**Comment by lewis6991** [comment]
Ah that makes sense. Not sure why I didn't realise that when I wrote the code originally.

If it works without then go ahead!

**Comment by tomtomjhj** [comment]
Oh I got a problem with that approach. There is infinite recursion `get_folds_levels` → `parse` → `_do_callback` → `on_changedtree` → `get_folds_levels`.

https://github.com/neovim/neovim/blob/2d40f5e843a99a67efd55304d17ccd48d47d9013/runtime/lua/vim/treesitter/languagetree.lua#L305-L312

`self._trees` is updated after `'changedtree'` callback, so `self._parser.parse` keeps returning the same `tree_changes`.

**Comment by lewis6991** [comment]
Ah that's why I didn't include `parse` originally, it should be handled outside, we may need to revert that change and rethink this.

**Comment by lewis6991** [comment]
Can we move `parse` to `foldexpr()`?

When `get_fold_levels()` is invoked in the langtree callbacks it should already be valid, but since we're scheduling it may become invalid which is why I originally had the guard in there.

`on_bytes` should handle insertions and deletions so the indices in `FoldLevels` should remain in sync.

**Comment by tomtomjhj** [comment]
I'm in favor of allowing the recursion (by swapping line 311 and 312), because it just looks more natural. The only code on github I can find that depends on the current behavior (i.e. `on_changedtree` callback accessing the old tree with `LanguageTree:trees` is this: https://github.com/HampusHauffman/bionic.nvim/blob/3523383b6db41046a3d813ac65dd57e78e61babb/lua/bionic/init.lua#L62 (found from https://github.com/search?q=on_changedtree+trees%28%29+language%3ALua+NOT+is%3Afork+NOT+path%3Aruntime%2Flua&type=code`)

**Comment by tomtomjhj** [comment]
> Can we move parse to foldexpr()?

Oh that sounds good. I'll try.

**Comment by tomtomjhj** [comment]
If I understood correctly, moving `parse` to `foldexpr()` doesn't work, because in the `new_row > old_row` case of `on_bytes`, the fold level of the additional lines are not defined (just as #24167). So `parse` + `get_folds_levels` is necessary.

Also, I just noticed that `vim.schedule` is necessary anyways due to the textlock. So there is no recursion involving on_changedtree in the current version.

So I propose to keep `parse` in `get_fold_levels` and fix `recompute_folds` only.

**Comment by lewis6991** [comment]
Ok let's go with that.

I think we need some more tests to nail this down.

**Comment by tomtomjhj** [comment]
added some tests

**Review by lewis6991 — APPROVED** [review]
Looks great! Thanks for all the hard work!

Let me know when to merge.

**Comment by tomtomjhj** [comment]
can be merged now

**Pull request merged**